### PR TITLE
165 scene/state

### PR DESCRIPTION
**Pull request by kaitlyn-sharo**
Created a script that will generate two api files from the swagger api. The first is a slightly modified copy of the swagger api that has some removed properties and a few more required properties. The second is a state_change api that removes many of the required properties from scenes/state. This PR also incorporates that state_change file into the validation. To test:

```
python3 validator.py -f test.yaml --update
```

Message from Darren that this PR is based on.

**API vs. Validator required differences:**
- Scenario validator requires Scenario.scenes; API does not;
- Scenario validator requires Action.probe_id and Action.choice; API does not;
- Scenario validator requires all Vitals properties to be specified; API does not.
- Valid values for InjuryStatusEnum should only be hidden, discoverable, or visible.  The others are only used at runtime by the server.

**There are a few properties that are allowed in the API but not in the validator:**
- Action.justification
- Character.visited
- Scenario.session_complete
- State.scenario_complete and State.elapsed_time
- NOTE: Scenario.start_time (wall clock local start time) used to exist in the API, but as it's not clear it's needed or used, I was thinking to remove it...

**When validating Scenes.state:**
- Only State.unstructured is required in State
- If State.mission is included, then only Mission.unstructured is required
- If State.environment is included, then Environment.sim_environment isn't required (i.e., nothing is)
- If Environment.sim_environment is included, then SimEnvironment.unstructured should be required (and type should be forbidden, but that's for later)
- If Environment.decision_environment is included, then DecisionEnvironment.unstructured should be required
- If DecisionEnvironment.aid_delay is included, then nothing should be required
- Other than that, validation is the same as for initial state.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -40,7 +40,14 @@ To run the validator, execute the following command:
 ```
 python3 validator.py -f [path_to_file]
 ```
-Ensure that the path leads to a yaml file
+Ensure that the path leads to a yaml file.
```
Comment (nextcen-dgemoets) [inline README.md:43]: The usage message makes it look like `-f` is optional, but it appears to be required.
`usage: validator.py [-h] [-f PATH] [-u]`

File: api_files/generator.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+        required_vitals.append('avpu')
+        required_vitals.append('mental_status')
+        required_vitals.append('breathing')
+        required_vitals.append('hrpmin')
+        required_vitals.append('Spo2')
+        new_api['components']['schemas']['Vitals']['required'] = required_vitals
+
+        # injury status enum should only include hidden, discoverable, or visible
+        new_api['components']['schemas']['InjuryStatusEnum']['enum'] = ['hidden', 'discoverable', 'visible']
+
```
Comment (nextcen-dgemoets) [inline api_files/generator.py:74]: Can we add a new rule I just thought of:
- `restricted_actions` cannot include `TAG_CHARACTER` and `END_SCENARIO`.  I don't want them to have a situation where they've set `end_scenario_allowed` to true but `restricted_actions` to include `END_SCENARIO`.

**Review by nextcen-dgemoets — CHANGES_REQUESTED** [review]
Everything looks super-solid.  All of the rules seem to work.  One very minor thing and a new rule to add, but otherwise ready to go.

I did have one more question-- there's an argument not to check in `state_changes.yaml` into source control because it's a generated file.  But I don't think there's anything wrong with checking it in, and prevents you from having to generate it the first time around, especially since there's no dependency set up where running the validator will generate the file if it's not already there.  So I think we're good-- I just wanted to mention it.

File: README.md (hunk 1 above)
Comment (kaitlyn-sharo) [reply README.md:43]: It is not required. When we discussed this yesterday, I decided to change it. You should just be able to run `validator.py -u` now.

Comment (nextcen-dgemoets) [reply README.md:43]: But if I'm reading the usage right, it's saying everything is optional.  If you don't use `-u`, then `-f` is required, right?

**Comment by kaitlyn-sharo** [comment]
> Everything looks super-solid. All of the rules seem to work. One very minor thing and a new rule to add, but otherwise ready to go.
> 
> I did have one more question-- there's an argument not to check in `state_changes.yaml` into source control because it's a generated file. But I don't think there's anything wrong with checking it in, and prevents you from having to generate it the first time around, especially since there's no dependency set up where running the validator will generate the file if it's not already there. So I think we're good-- I just wanted to mention it.

Yes, that's a good point about `state_changes.yaml`. I think that since we're going to be handing this to TA1, we want to reduce the amount of steps we hand them, so it would probably be nicest to just keep the two generated yaml files there so that they don't have to run it every time or worry about additional switches.

File: README.md (hunk 1 above)
Comment (kaitlyn-sharo) [reply README.md:43]: Right. How do I write that?

Comment (nextcen-dgemoets) [reply README.md:43]: Hmm, I think it's:
`usage: validator.py [-h] [-u -f PATH] | [-f PATH] ]`

So you either specify `-u` with `-f PATH`, or you just optionally specify `-f PATH`.

Comment (kaitlyn-sharo) [reply README.md:43]: Ah interesting, okay! I will fix that and see about adding the new rule.

File: api_files/generator.py (hunk 2 above)
Comment (kaitlyn-sharo) [reply api_files/generator.py:74]: So do you want this to always be true? Or do you want me to check if they set end_scenario_allowed to true and also restricted the actions?

Comment (nextcen-dgemoets) [reply api_files/generator.py:74]: It should always be true for `restricted_actions`.  It's limiting the allowable controlled vocab for this field.  If that makes it special-case enough that it needs to be in another ticket, then that's a-ok.  I don't think you need to cross-reference with `end_scenario_allowed`.

Comment (kaitlyn-sharo) [reply api_files/generator.py:74]: I just threw some code in that creates a new enum for restricted_actions based on actionTypeEnum, should work now!

**Review by nextcen-dgemoets — APPROVED** [review]
Everything has been addressed.

**Review by brianpippin — APPROVED** [review]

**Pull request merged**